# Logarithmic and Quadratic functional form (Example )

### Intro and objectives
#### Understand the application of logarithmic and quadratic functional forms in linear regression models

### In this lab you will learn:
1. examples of quadratic functional forms applied to simple regression models.
2. how to fit advanced regression models in Python.


## What I hope you'll get out of this lab
* Fit more advanced linear regression models
* How to interpret the results obtained

In [1]:
!pip install wooldridge
import wooldridge as woo
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 5.1 MB 13.2 MB/s 


# Example 1. Housing prices and air pollution

#### For a sample of 506 communities in the Boston area, we estimate a model relating median housing price (price) in the community to various community characteristics: nox is the amount of nitrogen oxide in the air, in parts per million; dist is a weighted distance of the community from five employment centers, in miles; rooms is the average number of rooms in houses in the community; and stratio is the average student-teacher ratio of schools in the community.



#### To study the relationship between housing prices and external factors , we postulate the following model:

$ log(price)=\beta_0+\beta_1*log(nox)+\beta_2*log(dist)+\beta_3*rooms+\beta_4*stratio+u $



### Using the data in HPRICE2 where n=506 individuals

In [23]:
hprice2 = woo.dataWoo('hprice2')


In [3]:
hprice2.head()

wage  educ  exper  tenure  nonwhite  female  married  numdep  smsa  \
0  3.10    11      2       0         0       1        0       2     1   
1  3.24    12     22       2         0       1        1       3     1   
2  3.00    11      2       0         0       0        0       2     0   
3  6.00     8     44      28         0       0        1       0     1   
4  5.30    12      7       2         0       0        1       1     0   

   northcen  ...  trcommpu  trade  services  profserv  profocc  clerocc  \
0         0  ...         0      0         0         0        0        0   
1         0  ...         0      0         1         0        0        0   
2         0  ...         0      1         0         0        0        0   
3         0  ...         0      0         0         0        0        1   
4         0  ...         0      0         0         0        0        0   

   servocc     lwage  expersq  tenursq  
0        0  1.131402        4        0  
1        1  1.175573      484        4  
2        0  1.098612        4        0  
3        0  1.791759     1936      784  
4        0  1.667707       49        4  

[5 rows x 24 columns]

In [24]:
hprice2.describe()

price       crime         nox       rooms        dist  \
count    506.000000  506.000000  506.000000  506.000000  506.000000   
mean   22511.509881    3.611536    5.549783    6.284051    3.795751   
std     9208.856171    8.590247    1.158395    0.702594    2.106137   
min     5000.000000    0.006000    3.850000    3.560000    1.130000   
25%    16850.000000    0.082000    4.490000    5.882500    2.100000   
50%    21200.000000    0.256500    5.380000    6.210000    3.210000   
75%    24999.000000    3.677000    6.240000    6.620000    5.187500   
max    50001.000000   88.975998    8.710000    8.780000   12.130000   

           radial     proptax     stratio     lowstat      lprice        lnox  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     9.549407   40.823715   18.459289   12.701482    9.941057    1.693091   
std      8.707259   16.853711    2.165820    7.238066    0.409255    0.201410   
min      1.000000   18.700001   12.600000    1.730000    8.517193    1.348073   
25%      4.000000   27.900000   17.400000    6.922500    9.732093    1.501853   
50%      5.000000   33.000000   19.100000   11.360000    9.961757    1.682688   
75%     24.000000   66.599998   20.200001   17.057500   10.126591    1.830980   
max     24.000000   71.099998   22.000000   39.070000   10.819798    2.164472   

         lproptax  
count  506.000000  
mean     5.931405  
std      0.396367  
min      5.231109  
25%      5.631212  
50%      5.799093  
75%      6.501290  
max      6.566672

In [5]:
type(hprice2)

pandas.core.frame.DataFrame

In [26]:
# We impose a simple, linear, model: 
# We specify CeoSalaries as the empirical dataset

reg = smf.ols(formula='np.log(price) ~ np.log(nox)+np.log(dist)+rooms+np.power(rooms,2)+stratio', data=hprice2)

In [27]:
# We fit the model
results = reg.fit()


In [28]:
b = results.params
print(f'b: \n{b}\n')

b: 
Intercept             13.385477
np.log(nox)           -0.901682
np.log(dist)          -0.086781
rooms                 -0.545113
np.power(rooms, 2)     0.062261
stratio               -0.047590
dtype: float64



## Based on the previous we have fitted the following model:

$ log(price)=13.39-0.902*log(nox)-0.087*log(dist)-0.545*rooms+0.062*rooms^2-0.047stratio+u $


## Let's compute t-tests of statistical significance and F-test of overall statistical significance

In [30]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:          np.log(price)   R-squared:                       0.603
Model:                            OLS   Adj. R-squared:                  0.599
Method:                 Least Squares   F-statistic:                     151.8
Date:                Fri, 18 Nov 2022   Prob (F-statistic):           7.89e-98
Time:                        16:49:28   Log-Likelihood:                -31.806
No. Observations:                 506   AIC:                             75.61
Df Residuals:                     500   BIC:                             101.0
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept             13.3855      0

### The F-statistic of overall significance (151.8) and its low p-value (7.89e-98) means that the model is significant.

### According to the t-statics and related p-values, all explanatory variables are significant.
### The R-squared is moderately large (0.603) specially for an econometric dataset.

## How do we interpret the equation?

#### This estimated equation implies that nox, distance to employment centers and student teacher ratio have negative effects on house prices.

#### 1% increase in nox decreases house prices by 0.902%

#### A unit increase in the student/teacher ratio (stratio) decreases house prices by 4.7%.

#### An increase in rooms from five to six increases price by about -54.5+12.4(5)=7.5%
#### An increase in rooms from six to seven increases price by about -54.5+12.4(6)=19.9%. This is a very strong increasing effect.

# Feel free to improve the model by adding additional regressors and other functional forms